content/hoon/lagoon/conv.md

```diff
@@ -0,0 +1,962 @@
+  |=  [shape=(list @) ind=@]
+  =/  shape  (flop shape)
+  =/  i=@  0
+  =|  res=(list @)
+  ?>  (^lth ind (roll shape ^mul))
+  |-  ^-  (list @)
+  ?:  (^gte i (lent shape))  (flop res)
+  %=    $
+    res  `(list @)`(snoc res (^mod ind (snag i shape)))
+    ind  (^div ind (snag i shape))
+    i    (^add i 1)
+  ==
+```
+
+---
+
+## `+get-item-index` {#getitemindex}
+
+Utility function to get the item index from a multi-dimensional index.
+
+#### Accepts
+
+Parameters as specified in source
+
+#### Produces
+
+An index `@`.
+
+#### Source
+
+```hoon
+++  get-item-index
+  |=  [shape=(list @) num=@]
+  ^-  @
+  =/  len  (roll shape ^mul)
+  =-  (roll - ^add)
+  ^-  (list @)
+  %+  turn  shape
+  |=  wid=@
+  (^mod (^div len wid) num)
+```
+
+---
+
+## `+ravel` {#ravel}
+
+Utility function to flatten a `$ray` into a list of atoms.
+
+#### Accepts
+
+A `$ray`.
+
+#### Produces
+
+A `(list @)`.
+
+#### Source
+
+```hoon
+++  ravel
+  |=  a=ray
+  ^-  (list @)
+  (snip (rip bloq.meta.a data.a))
+```
+
+---
+
+## `+en-ray` {#enray}
+
+Convert a `$baum` to a `$ray`.
+
+#### Accepts
+
+A `$baum` (or unwrapped `$ray`).
+
+#### Produces
+
+A `$ray`.
+
+#### Source
+
+```hoon
+++  en-ray    :: baum to ray
+  |=  =baum
+  ^-  ray
+  =/  a=ray  (zeros meta.baum)
+  =/  i  0
+  =/  n  (roll shape.meta.a ^mul)
+  |-
+  ?:  =(i n)  a
+  %=  $
+    i  +(i)
+    data.a
+      %+  con
+        data.a
+      %+  lsh
+        [bloq.meta.a i]
+      (get-item-baum baum (get-dim shape.meta.a i))
+  ==
+```
+
+---
+
+## `+de-ray` {#deray}
+
+Utility function to convert a `$ray` to a `$baum`.
+
+#### Accepts
+
+A `$ray`.
+
+#### Produces
+
+A `$baum` (or unwrapped `$ray`).
+
+#### Source
+
+```hoon
+++  de-ray    :: ray to baum
+  |=  =ray
+  ^-  baum
+  |^
+  :-  meta.ray
+  ^-  ndray
+  ::
+  =,  meta.ray
+  ?:  =(1 (lent shape))
+    ::  Snip off tail which is the pinned 0x1 MSB
+    (snip (rip bloq data.ray))
+  ::
+  ?:  =(2 (lent shape))
+    =/  dims  (flop shape)
+    =|  fin=(list ndray)
+    =|  els=ndray
+    |-
+    ?:  =(0x1 data.ray)  (welp ;;((list ndray) fin) ~[;;((list ndray) els)])
+    %=  $
+      els   (snip (rip bloq (cut bloq [0 +((snag 0 dims))] data:(spac `^ray`[[~[(snag 0 dims) 1] bloq kind tail] `@ux`data.ray]))))
+      fin   ?~  els  fin  :: skip on first row
+            ?~  fin  `(list (list ndray))`~[;;((list ndray) els)]
+            (welp ;;((list (list ndray)) fin) ~[;;((list ndray) els)])
+      data.ray  (rsh [bloq (snag 0 dims)] data.ray)
+    ==
+  !!
+```
+
+---
+
+## `+rip` {#rip}
+
+Utility function to recursively rip a bite into a list of atoms.
+
+#### Accepts
+
+Parameters as specified in source
+
+#### Produces
+
+A `(list @)`.
+
+#### Source
+
+```hoon
+++  rip
+  |=  [a=bite b=@]
+  ^-  (list @)
+  =/  cnt  ?@(a 1 +.a)
+  ?~  (^rip a b)  (reap cnt 0)
+  (^rip a b)
+--
+```
+
+---
+
+## `+check` {#check}
+
+Check if a `$ray` is valid according to its metadata.
+
+#### Accepts
+
+A `$ray`.
+
+#### Produces
+
+A loobean.
```

Comment:
I'd prefer "A `$flag`." for these Produces sections. This only occurs once in this PR, the other mentions of "loobean" are fine.